File: packages/akilent-php/tests/ClientTest.php
<?php

declare(strict_types=1);

namespace Akilent\Tests;

use Akilent\Client;
use Akilent\ConflictError;
use Akilent\NotFoundError;
use Akilent\SignatureVerificationError;
use Akilent\Webhooks;
use PHPUnit\Framework\TestCase;

final class ClientTest extends TestCase
{
    /** @var array<int,array{method:string,url:string,headers:array<string,string>,body:?string}> */
    private array $calls = [];

    private function client(callable $responder): Client
    {
        $c = new Client('ak_test_abc', 'https://akilent.test');
        $c->transport->sender = function (string $method, string $url, array $headers, ?string $body) use ($responder): array {
            $this->calls[] = compact('method', 'url', 'headers', 'body');
            return $responder($method, $url, $headers, $body);
        };
        return $c;
    }

    public function testSendPostsWithAutoIdempotencyKey(): void
    {
        $c = $this->client(fn () => [202, json_encode(['id' => 1, 'public_id' => 'msg_1', 'status' => 'queued']), []]);
        $out = $c->messages->send(from: 'a@x.com', to: 'b@y.com', text: 'hi');

        $this->assertSame('msg_1', $out['public_id']);
        $this->assertSame('POST', $this->calls[0]['method']);
        $this->assertSame('https://akilent.test/api/v1/messages', $this->calls[0]['url']);
        $this->assertSame('Bearer ak_test_abc', $this->calls[0]['headers']['Authorization']);
        $this->assertArrayHasKey('Idempotency-Key', $this->calls[0]['headers']);
        $this->assertStringContainsString('"from":"a@x.com"', $this->calls[0]['body']);
    }

    public function testListBuildsQueryString(): void
    {
        $c = $this->client(fn () => [200, json_encode(['data' => [], 'total' => 0]), []]);
        $c->messages->list(['status' => 'delivered', 'limit' => 5]);
        $this->assertStringContainsString('status=delivered', $this->calls[0]['url']);
        $this->assertStringContainsString('limit=5', $this->calls[0]['url']);
    }

    public function testNotFoundRaisesTypedError(): void
    {
        $c = $this->client(fn () => [404, json_encode(['error' => ['code' => 'not_found', 'message' => 'nope', 'request_id' => 'req_9']]), []]);
        try {
            $c->messages->retrieve('msg_missing');
            $this->fail('expected NotFoundError');
        } catch (NotFoundError $e) {
            $this->assertSame(404, $e->statusCode);
            $this->assertSame('not_found', $e->errorCode);
            $this->assertSame('req_9', $e->requestId);
        }
    }

    public function testConflictOnIdempotencyReuse(): void
    {
        $c = $this->client(fn () => [409, json_encode(['error' => ['code' => 'idempotency_key_reuse', 'message' => 'reused']]), []]);
        $this->expectException(ConflictError::class);
        $c->messages->send(from: 'a@x.com', to: 'b@y.com', text: 'hi', idempotencyKey: 'k1');
    }

    public function testRetriesOn503ThenSucceeds(): void
    {
        $n = 0;
        $c = $this->client(function () use (&$n): array {
            $n++;
            return $n < 2 ? [503, '', []] : [200, json_encode(['data' => []]), []];
        });
        $c->messages->list();
        $this->assertSame(2, $n);
    }

    public function testIteratePagesThroughResults(): void
    {
        $pages = [
            [200, json_encode(['data' => [['id' => 'm1'], ['id' => 'm2']], 'total' => 3]), []],
            [200, json_encode(['data' => [['id' => 'm3']], 'total' => 3]), []],
        ];
        $i = 0;
        $c = $this->client(function () use (&$i, $pages): array {
            return $pages[$i++];
        });
        $ids = [];
        foreach ($c->messages->iterate([], 2) as $row) {
            $ids[] = $row['id'];
        }
        $this->assertSame(['m1', 'm2', 'm3'], $ids);
    }

    public function testWebhookVerifyRoundTrips(): void
    {
        $secret = 'whsec_test';
        $body = '{"event":"message.sent"}';
        $ts = time();
        $sig = "t={$ts},v1=" . hash_hmac('sha256', "{$ts}.{$body}", $secret);
        $this->assertTrue(Webhooks::verify($body, $sig, $secret));

        $this->expectException(SignatureVerificationError::class);
        Webhooks::verify($body . 'x', $sig, $secret);
    }
}

File: packages/akilent-go/client_test.go
package akilent

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"net/http"
	"net/http/httptest"
	"strconv"
	"sync/atomic"
	"testing"
	"time"
)

func newTestClient(h http.HandlerFunc) (*Client, *httptest.Server) {
	srv := httptest.NewServer(h)
	c := NewWithConfig(Config{APIKey: "ak_test_abc", BaseURL: srv.URL})
	return c, srv
}

func TestSendPostsWithAutoIdempotencyKey(t *testing.T) {
	var gotAuth, gotIdem, gotBody string
	c, srv := newTestClient(func(w http.ResponseWriter, r *http.Request) {
		gotAuth = r.Header.Get("Authorization")
		gotIdem = r.Header.Get("Idempotency-Key")
		buf := make([]byte, r.ContentLength)
		r.Body.Read(buf)
		gotBody = string(buf)
		w.WriteHeader(202)
		fmt.Fprint(w, `{"id":1,"public_id":"msg_1","status":"queued"}`)
	})
	defer srv.Close()

	msg, err := c.Messages.Send(context.Background(), SendParams{From: "a@x.com", To: "b@y.com", Text: "hi"})
	if err != nil {
		t.Fatal(err)
	}
	if msg.PublicID != "msg_1" {
		t.Fatalf("public_id = %q", msg.PublicID)
	}
	if gotAuth != "Bearer ak_test_abc" {
		t.Fatalf("auth = %q", gotAuth)
	}
	if gotIdem == "" {
		t.Fatal("missing Idempotency-Key")
	}
	if want := `"from":"a@x.com"`; !contains(gotBody, want) {
		t.Fatalf("body %q missing %q", gotBody, want)
	}
}

func TestListSendsQueryParams(t *testing.T) {
	var gotQuery string
	c, srv := newTestClient(func(w http.ResponseWriter, r *http.Request) {
		gotQuery = r.URL.RawQuery
		fmt.Fprint(w, `{"data":[],"total":0}`)
	})
	defer srv.Close()

	_, err := c.Messages.List(context.Background(), map[string]string{"status": "delivered"})
	if err != nil {
		t.Fatal(err)
	}
	if !contains(gotQuery, "status=delivered") {
		t.Fatalf("query = %q", gotQuery)
	}
}

func TestNotFoundReturnsTypedError(t *testing.T) {
	c, srv := newTestClient(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(404)
		fmt.Fprint(w, `{"error":{"code":"not_found","message":"nope","request_id":"req_9"}}`)
	})
	defer srv.Close()

	_, err := c.Messages.Get(context.Background(), "msg_missing")
	if !IsNotFound(err) {
		t.Fatalf("expected IsNotFound, got %v", err)
	}
	ae := err.(*APIError)
	if ae.Code != "not_found" || ae.RequestID != "req_9" {
		t.Fatalf("bad APIError: %+v", ae)
	}
}

func TestConflictOnIdempotencyReuse(t *testing.T) {
	c, srv := newTestClient(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(409)
		fmt.Fprint(w, `{"error":{"code":"idempotency_key_reuse","message":"reused"}}`)
	})
	defer srv.Close()

	_, err := c.Messages.Send(context.Background(), SendParams{From: "a@x.com", To: "b@y.com", Text: "hi", IdempotencyKey: "k1"})
	if !IsConflict(err) {
		t.Fatalf("expected IsConflict, got %v", err)
	}
}

func TestRetriesOn503(t *testing.T) {
	var calls int32
	c, srv := newTestClient(func(w http.ResponseWriter, r *http.Request) {
		if atomic.AddInt32(&calls, 1) < 2 {
			w.WriteHeader(503)
			return
		}
		fmt.Fprint(w, `{"data":[]}`)
	})
	defer srv.Close()

	if _, err := c.Messages.List(context.Background(), nil); err != nil {
		t.Fatal(err)
	}
	if calls != 2 {
		t.Fatalf("calls = %d, want 2", calls)
	}
}

func TestVerifyWebhookSignature(t *testing.T) {
	secret := "whsec_test"
	body := []byte(`{"event":"message.sent"}`)
	ts := strconv.FormatInt(time.Now().Unix(), 10)
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(ts + "."))
	mac.Write(body)
	header := "t=" + ts + ",v1=" + hex.EncodeToString(mac.Sum(nil))

	if err := VerifyWebhookSignature(body, header, secret); err != nil {
		t.Fatalf("verify failed: %v", err)
	}
	if err := VerifyWebhookSignature(append(body, 'x'), header, secret); err == nil {
		t.Fatal("expected verification failure on tampered body")
	}
}

func contains(s, sub string) bool {
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}
